apps/tap/test/tap.js

```diff
@@ -62,64 +85,328 @@ contract('Tap app', accounts => {
+          async () =>
+            await tap.addTokenTap(token.address, tapAmt, { from: unauthorized })
+        )
+      })
+    })
+  })
+
+  context("> removeTokenTap", () => {
+    context('sender has REMOVE_TOKEN_TAP_ROLE', () => {
+      context('and token is ETH or ERC20', () => {
+        context('and token does exist in mapping', () => {
+          it('it should remove tap token in mapping', async () => {
+            token2 = await TokenMock.new(authorized, 10000)
+
+            await tap.addTokenTap(ETH, tapAmt, { from: authorized })
+            await tap.addTokenTap(token2.address, tapAmt, { from: authorized })
+
+            await tap.removeTokenTap(ETH, { from: authorized })
+            const receipt = await tap.removeTokenTap(token2.address, { from: authorized })
+            assertEvent(receipt, 'RemoveTokenTap')
+          })
+        })
+        context('but token does not exist in mapping', () => {
+          it('it should revert', async () => {
+            const token = await TokenMock.new(authorized, 10000)
+            await assertRevert(
+              async () =>
+                await tap.removeTokenTap(token.address, {
+                  from: authorized
+                })
+            )
+          })
+        })
+      })
+      context('but token is not ETH or ERC20', () => {
+        it('it should revert', async () => {
+          await assertRevert(
+            async () =>
+              await tap.removeTokenTap(root, { from: authorized })
+          )
+        })
+      })
+    })
+    context('sender does not have REMOVE_TOKEN_TAP_ROLE', () => {
+      it('it should revert', async () => {
+        const token = await TokenMock.new(authorized, 10000)
+        await tap.addTokenTap(token.address, tapAmt, { from: authorized })
+
+        await assertRevert(
+          async () =>
+            await tap.removeTokenTap(token.address, { from: unauthorized })
+        )
+      })
+    })
+  })
+
+  context("updateTokenTap", () => {
+    context("sender does have UPDATE_TOKEN_TAP_ROLE", () => {
+      context('and token is ETH or ERC20', () => {
+        it("it should update tap rate if within monthly limit", async () => {
+          const token = await TokenMock.new(authorized, 10000)
+          await tap.addTokenTap(ETH, tapAmt, { from: authorized })
+          await tap.addTokenTap(token.address, tapAmt, { from: authorized })
+
+          await tap.updateTokenTap(ETH, 10, { from: authorized })
+          await tap.updateTokenTap(token.address, 10, { from: authorized })
+         // assertEvent(receipt, 'UpdateTokenTap')
+
+        })
+        it("it should revert if tap rate exceeds limit within 30 days", async () => {
+          //I'd like to submit a transaction with a spoofed date?
+          const token = await TokenMock.new(authorized, 10000)
+          await tap.addTokenTap(ETH, tapAmt, { from: authorized })
+          await tap.addTokenTap(token.address, tapAmt, { from: authorized })
+
```

Comment:
Here I'd like to test two cases:

1. When it exceeds the monthly tap rate, it should revert
2. When it exceeds the monthly tap rate, but 30 days has passed since the last tap update it should continue.

Best way to spoof a timestamp on a block?